mailbox/plugin/rspamd/src/main/java/org/apache/james/mailbox/rspamd/RSpamDListener.java

```diff
@@ -0,0 +1,206 @@
+
+    }
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(RSpamDListener.class);
+
+    private static final int LIMIT = 1;
+    private static final Group GROUP = new RSpamDListenerGroup();
+
+    private final RSpamDHttpClient rSpamDHttpClient;
+    private final MailboxManager mailboxManager;
+    private final MailboxSessionMapperFactory mapperFactory;
+    private final SystemMailboxesProvider systemMailboxesProvider;
+
+    @Inject
+    public RSpamDListener(RSpamDHttpClient rSpamDHttpClient, MailboxManager mailboxManager, MailboxSessionMapperFactory mapperFactory, SystemMailboxesProvider systemMailboxesProvider) {
+        this.rSpamDHttpClient = rSpamDHttpClient;
+        this.mailboxManager = mailboxManager;
+        this.mapperFactory = mapperFactory;
+        this.systemMailboxesProvider = systemMailboxesProvider;
+    }
+
+    @Override
+    public Group getDefaultGroup() {
+        return GROUP;
+    }
+
+    @Override
+    public boolean isHandling(Event event) {
+        return event instanceof MessageMoveEvent || event instanceof MailboxEvents.Added;
+    }
+
+    @Override
+    public Publisher<Void> reactiveEvent(Event event) {
+        if (event instanceof MessageMoveEvent) {
+            return handleMessageMoved((MessageMoveEvent) event);
+        } else if (event instanceof MailboxEvents.Added) {
+            return handleMessageAdded((MailboxEvents.Added) event);
+        }
+        return Mono.empty();
+    }
+
+    private Mono<Void> handleMessageAdded(MailboxEvents.Added addedEvent) {
+        return isAppendedToInbox(addedEvent)
+            .filter(FunctionalUtils.identityPredicate())
+            .doOnNext(isHam -> LOGGER.debug("Ham event detected, EventId = {}", addedEvent.getEventId().getId()))
+            .flatMap(any -> reportHamWhenAdded(addedEvent, mailboxManager.createSystemSession(Username.of(getClass().getCanonicalName()))));
+    }
+
+    private Mono<Void> handleMessageMoved(MessageMoveEvent messageMoveEvent) {
+        return handleMessageMoved(retrieveHeaderAndContentInputStreamsFromMovedMessage(messageMoveEvent), messageMoveEvent);
+    }
+
+    private Mono<Void> reportHamWhenAdded(MailboxEvents.Added addedEvent, MailboxSession session) {
+        return mapperFactory.getMailboxMapper(session)
+            .findMailboxById(addedEvent.getMailboxId())
+            .map(mailbox -> Pair.of(mailbox, mapperFactory.getMessageMapper(session)))
+            .flatMapMany(pair -> Flux.fromIterable(MessageRange.toRanges(addedEvent.getUids()))
+                .flatMap(range -> pair.getRight().findInMailboxReactive(pair.getLeft(), range, MessageMapper.FetchType.FULL, LIMIT)))
+            .map(Throwing.function(MailboxMessage::getFullContent))
+            .map(this::getHeaderAndContentInputStream)
+            .flatMap(headerAndContent -> rSpamDHttpClient.reportAsHam(headerAndContent.getLeft(), headerAndContent.getRight()), ReactorUtils.DEFAULT_CONCURRENCY)
```

Comment:
IMO all of this code would be way easier if we can `rSpamDHttpClient.reportAsHam(fullContent)` and `rSpamDHttpClient.reportAsSpam(fullContent)`.

Now we are parsing the mime message to split header and content to merge it again later down the road, which is complex and inefficient.

Can we do this refactoring please?

Reply:
CF https://github.com/apache/james-project/pull/1100